File: drouter-plugin/src/main/groovy/com/didi/drouter/generator/ClassClassify.java
package com.didi.drouter.generator;

import com.didi.drouter.plugin.RouterSetting;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import javassist.ClassPool;
import javassist.CtClass;

/**
 * Created by gaowei on 2018/8/30
 */
public class ClassClassify {

    private List<AbsRouterCollect> classifies = new ArrayList<>();

    public ClassClassify(ClassPool pool, RouterSetting setting) {
        classifies.add(new RouterCollect(pool, setting));
        classifies.add(new ServiceCollect(pool, setting));
        classifies.add(new InterceptorCollect(pool, setting));
    }

    public boolean doClassify(CtClass ct) {
        boolean take = false;
        for (int i = 0; i < classifies.size(); i++) {
            AbsRouterCollect cf = classifies.get(i);
            take = cf.collect(ct) || take;
        }
        return take;
    }

    public void generatorRouter(File routerDir) throws Exception {
        for (int i = 0; i < classifies.size(); i++) {
            AbsRouterCollect cf = classifies.get(i);
            cf.generate(routerDir);
        }
    }

}

File: plugins/drouter/src/main/kotlin/com/didi/drouter/generator/ClassClassify.kt
package com.didi.drouter.generator

import com.didi.drouter.plugin.RouterSetting.Parse
import com.didi.drouter.utils.Logger
import javassist.ClassPool
import javassist.CtClass
import java.io.File

/**
 * Created by gaowei on 2018/8/30
 */
class ClassClassify(pool: ClassPool, setting: Parse) {
    private val classifies: MutableList<AbsRouterCollect> = ArrayList()

    init {
        classifies.add(RouterCollect(pool, setting))
        classifies.add(ServiceCollect(pool, setting))
        classifies.add(InterceptorCollect(pool, setting))
    }

    fun doClassify(ct: CtClass): Boolean {
        var take = false
        for (i in classifies.indices) {
            val cf = classifies[i]
            take = cf.collect(ct) || take
        }
        if (take) {
            Logger.d("    == router class: " + ct.name)
        }
        return take
    }

    @Throws(Exception::class)
    fun generatorRouter(routerDir: File) {
        for (i in classifies.indices) {
            val cf = classifies[i]
            cf.generate(routerDir)
        }
    }
}
